src/dune_rules/mdx.ml

```diff
@@ -68,22 +68,48 @@ module Deps = struct
       [ A "deps"; Dep (Path.build files.Files.src) ]
       ~stdout_to:files.Files.deps
 
+  let path_escapes_dir str =
+    match Path.Local.of_string_result str with
+    | Ok _ -> false
+    | Error `Outside_the_workspace -> true
+
   let to_path ~dir str =
-    let local = Path.Local.of_string str in
-    let build = Path.Build.append_local dir local in
-    Path.build build
+    if not (Filename.is_relative str) then Error (`Absolute str)
+    else
+      let path = Path.relative_to_source_in_build_or_external ~dir str in
+      if not (Path.is_managed path) then Error (`Escapes_workspace str)
```

Comment:
It might be easier to just pattern match on `path` directly. You'd avoid the `is_managed` check and `as_in_build_dir_exn` in one go. I.e.

```
match path with
| In_build_dir .. -> Ok path
| _ -> (* error detection goes here *)
```